api/lib/reducers/polkadotV0-reducers.js

```diff
@@ -428,30 +434,24 @@ function stakeDetailsReducer(network, message, reducers) {
   }
 }
 
-async function extractInvolvedAddresses(
+function extractInvolvedAddresses(
   lunieTransactionType,
   signer,
   message,
-  db
+  messageIndex,
+  blockEvents,
 ) {
   let involvedAddresses = []
   if (lunieTransactionType === lunieMessageTypes.SEND) {
     involvedAddresses = involvedAddresses.concat([signer, message.args[0]])
   } else if (lunieTransactionType === lunieMessageTypes.STAKE) {
     involvedAddresses = involvedAddresses.concat([signer], message.validators)
   } else if (lunieTransactionType === lunieMessageTypes.CLAIM_REWARDS) {
-    // the rewards claiming happens for all delegators so we get the rewards for the claimed era and extract all addresses from there
-    const dbRewards = (await db.getRewards(delegatorAddress)) || []
-    const involvedRewards = dbRewards.filter(({ height, validator }) => {
-      return height === String(message.args[1]) && validator === message.args[0]
-    })
-    involvedAddresses = involvedAddresses.concat(
-      [signer],
-      involvedRewards.map(({ address }) => address)
-    )
-    if (involvedRewards.length > 0) {
-      involvedAddresses = involvedAddresses.concat(involvedRewards[0].validator)
-    }
+    // we get all reward target addresses from block events, we don't include signer
+    // as sender address doesn't need to be a nominator or validator
+    involvedAddresses = blockEvents
+      .filter(({ phase }) => parseInt(phase.toHuman().ApplyExtrinsic) === messageIndex)
+      .map(({ event }) => event.toHuman().data[0])
```

Comment:
sweet!

Reply:
I'll test tomorrow again, but should work (I claimed all my rewards today :-)

Reply:
mmm now this is a mess, we should move our transaction splitting logic to be based on events instead of extrinsics.

This case is a good example I think:

Claim rewards frequently is a `utility.batch` multi tx, including several `staking.payoutStakers` tx for each validator/era tuple. We split the tx based on that but that doesn't make sense, those extrinsics will trigger multiple `staking.Reward` events that are better to get all the individual messages.

Btw I miss the semantic block returned by sidecar (events are attached to extrinsics for easy correlation).